### PR TITLE
[Refactoring] SR-6051 Expansion of `switch` statement missing cases

### PR DESCRIPTION
**Pull request by Kacper20**
Refactor is available when the cursor is placed on `switch` keyword. 
There are still a few more tests to be written so it's WIP.
Addresses [SR-6051](https://bugs.swift.org/browse/SR-6051).

cc @nkcsgexi

### CODE REVIEW

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -1800,6 +1800,48 @@ collectAvailableRefactoringsAtCursor(SourceFile *SF, unsigned Line,
...
+                                              EditorConsumerInsertStream &OS
+                                              ) {
+  // To find the subject enum decl for this switch statement; if failing,
+  // issue errors.
+  EnumDecl *SubjectED = nullptr;
+  if (auto SubjectTy = SwitchS->getSubjectExpr()->getType()) {
+    SubjectED = SubjectTy->getAnyNominal()->getAsEnumOrEnumExtensionContext();
+  }
+  if (!SubjectED) {
+    DiagEngine.diagnose(ExpandedStmtLoc, diag::no_subject_enum);
```
Comment (CodaFi) [inline lib/IDE/Refactoring.cpp]: This... isn’t correct.  For one you can switch over arbitrary expressions, including tuples, so `isApplicable` should probably be updated to take into account what expressions this action can handle.

Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: It's actually how Expand Default is implemented right now, before this PR is introduced. I've just moved the code and generalized it to add handling when `switch` is selected. You're right — I will fix `isApplicable` in both `Expand Default` and `Expand Switch Cases`. Thanks!

Comment (CodaFi) [reply lib/IDE/Refactoring.cpp]: Oh, sorry that wasn't your bug.  Syntactic expansion here is a very nuanced topic for me 😄

**Comment by Kacper20** [comment]
@nkcsgexi @CodaFi Currently in `Expand Default` implementation there are other places when applicability is handled in `performChange` method like that:
```
  if (UnhandledElements.empty()) {
    DiagEngine.diagnose(ExpandedStmtLoc, diag::no_remaining_cases);
    return true;
  }
```

What's the reason to do this? Should we move it also to the `isApplicable` so that the refactor will be grayed out in Xcode? Are there any advantages of handling this in `performChange`?

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: It's great that you've spotted this. There are couple other places like that in `Expand Default` `performChange` — where we could move the checking to `isApplicable`. It's a perfect moment to... refactor it :)

**Comment by nkcsgexi** [comment]
@Kacper20 I think moving it to `isApplicable` makes sense. Could you update it?

**Comment by Kacper20** [comment]
Yep, thanks for the answer! So let's wait with the review — I will update it tomorrow with the changes :)

**Comment by Kacper20** [comment]
@CodaFi I think it's ready to review — I've moved the checking to `isApplicable` and written couple more tests.

**Review by CodaFi — CHANGES_REQUESTED** [review]
Just one change request.  Everything else looks much improved.  Thanks!

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -633,6 +633,15 @@ class TokenBasedRefactoringAction : public RefactoringAction {
   }
 };
 
+struct ApplicabilityContext {
```
[the remark this answers is not in the document]
Comment (Kacper20) [reply lib/IDE/Refactoring.cpp]: Yeah, that's the reason. The problem is that action has `static isApplicable()` method and `isApplicable` which is not static calls static one and is not overridable. I've changed it but forgot to commit it. It'll now have access to the file through `ResolvedRangeInfo`, as discussed in: https://github.com/apple/swift/pull/12268

Thanks!

**Comment by Kacper20** [comment]
@CodaFi Thanks! Fixed now.

**Comment by Kacper20** [comment]
@CodaFi could you run the tests?

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Review by CodaFi — APPROVED** [review]

**Comment by Kacper20** [comment]
Could you merge it? @CodaFi @nkcsgexi Thanks in advance! :)

**Pull request merged**